File: shifai-ios/ShifAI/Data/PerformanceMonitor.swift
import Foundation
import os.signpost

/// Performance Monitor — lightweight instrumentation for production.
/// Tracks startup time, DB queries, sync duration, and screen transitions.
/// No third-party dependency.
final class PerformanceMonitor {

    static let shared = PerformanceMonitor()

    private let log = OSLog(subsystem: "com.shifai", category: "performance")
    private var marks: [String: CFAbsoluteTime] = [:]

    private init() {}

    // MARK: - Measurement

    func start(_ label: String) {
        marks[label] = CFAbsoluteTimeGetCurrent()
        os_signpost(.begin, log: log, name: "measure", "%{public}s", label)
    }

    func end(_ label: String) -> TimeInterval? {
        guard let startTime = marks.removeValue(forKey: label) else { return nil }
        let duration = CFAbsoluteTimeGetCurrent() - startTime
        os_signpost(.end, log: log, name: "measure", "%{public}s", label)

        // Log if slow
        let threshold = thresholdFor(label)
        if duration > threshold {
            CrashReporter.shared.log(
                .performanceBudgetExceeded,
                context: "\(label): \(String(format: "%.0f", duration * 1000))ms (budget: \(String(format: "%.0f", threshold * 1000))ms)"
            )
        }

        return duration
    }

    // MARK: - Convenience

    func measure<T>(_ label: String, block: () throws -> T) rethrows -> T {
        start(label)
        let result = try block()
        _ = end(label)
        return result
    }

    func measureAsync<T>(_ label: String, block: () async throws -> T) async rethrows -> T {
        start(label)
        let result = try await block()
        _ = end(label)
        return result
    }

    // MARK: - Budgets

    private func thresholdFor(_ label: String) -> TimeInterval {
        switch label {
        case "cold_start":          return 1.5
        case "warm_start":          return 0.5
        case "db_open":             return 0.2
        case "save_daily_log":      return 0.2
        case "load_chart":          return 0.3
        case "ml_prediction":       return 0.5
        case "pdf_generation":      return 3.0
        case "sync":                return 5.0
        case "tab_switch":          return 0.1
        case "widget_refresh":      return 1.0
        default:                    return 1.0
        }
    }
}

File: shifai-android/app/src/main/java/com/shifai/data/monitoring/PerformanceMonitor.kt
package com.shifai.data.monitoring

import android.os.SystemClock
import android.util.Log

/**
 * Performance Monitor — lightweight instrumentation for production.
 * Tracks startup time, DB queries, sync duration, and screen transitions.
 * Mirrors iOS PerformanceMonitor.swift.
 */
object PerformanceMonitor {

    private const val TAG = "PerfMonitor"
    private val marks = mutableMapOf<String, Long>()

    // ─── Measurement ───

    fun start(label: String) {
        marks[label] = SystemClock.elapsedRealtime()
    }

    fun end(label: String): Long? {
        val startTime = marks.remove(label) ?: return null
        val durationMs = SystemClock.elapsedRealtime() - startTime

        val budgetMs = budgetFor(label)
        if (durationMs > budgetMs) {
            Log.w(TAG, "⚠️ $label: ${durationMs}ms (budget: ${budgetMs}ms)")
            CrashReporter.log("PERF_EXCEEDED", "$label exceeded budget", "${durationMs}ms > ${budgetMs}ms")
        } else {
            Log.d(TAG, "✓ $label: ${durationMs}ms")
        }

        return durationMs
    }

    inline fun <T> measure(label: String, block: () -> T): T {
        start(label)
        val result = block()
        end(label)
        return result
    }

    suspend inline fun <T> measureAsync(label: String, block: () -> T): T {
        start(label)
        val result = block()
        end(label)
        return result
    }

    // ─── Budgets (from PERFORMANCE_BUDGET.md) ───

    private fun budgetFor(label: String): Long = when (label) {
        "cold_start" -> 1500
        "warm_start" -> 500
        "db_open" -> 200
        "save_daily_log" -> 200
        "load_chart" -> 300
        "ml_prediction" -> 500
        "pdf_generation" -> 3000
        "sync" -> 5000
        "tab_switch" -> 100
        "widget_refresh" -> 1000
        else -> 1000
    }
}
